### PR TITLE
Add support for onGlobalLayout

### PR DESCRIPTION
**Pull request by gamepro65**
ViewRootImpl is responsible for scheduling / traversing view callbacks with the Choreographer. In order to support onGlobalLayout we must use ViewRootImpl#setView with RenderSession's rootView to allow the proper callbacks to be triggered.  The problem is that #setView interacts with DisplayManager and WindowSession which are not fully setup by layout lib. So we can use ByteBuddy to stub out the integration with those classes and provide sensible defaults to allow access to this method.

### CODE REVIEW

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 1]
```diff
@@ -272,12 +280,21 @@ class Paparazzi(
...
+        val viewRootImpl = viewGroup.viewRootImpl
+        val params = sessionParamsBuilder.build()
+        viewRootImpl.setView(viewGroup.findRoot(), LayoutParams(params.screenWidth, params.screenHeight, 0, 0, 0), null)
+      }
     } catch (e: Exception) {
       throw RuntimeException(e)
     }
   }
 
+  private fun View.findRoot(): View {
```
Comment (JakeWharton) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:294]: super nit: +`tailrec`

You may have to pull the base case into its own conditional separate from the recursion case though

File: paparazzi/src/main/java/app/cash/paparazzi/internal/ServiceManagerInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package app.cash.paparazzi.internal
+
+import android.os.IBinder
+
+object ServiceManagerInterceptor {
+  @JvmStatic
+  fun intercept(
+    name: String
+  ): IBinder? {
+    if (name == "display") {
```
Comment (JakeWharton) [inline paparazzi/src/main/java/app/cash/paparazzi/internal/ServiceManagerInterceptor.kt:26]: could be a `when`, doesn't matter

**Review by JakeWharton — APPROVED** [review]
Wow this wiring is a pain in the ass. But if it works, 👍 .

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt (hunk 1 above)
Comment (gamepro65) [reply paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt]: 👍  Nice, ended up changing the compiled code into a while loop with no recursion

File: paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import android.hardware.display.IDisplayManager
+import android.hardware.display.IDisplayManagerCallback
+import android.hardware.display.IVirtualDisplayCallback
+import android.hardware.display.WifiDisplayStatus
+import android.media.projection.IMediaProjection
+import android.view.DisplayInfo
+import android.view.Surface
+
+class PaparazziDisplayManager : IDisplayManager.Stub() {
+
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt:32]:
```suggestion
```

File: paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  override fun getPreferredWideGamutColorSpaceId(): Int {
+    return 0
+  }
+
+  override fun registerCallback(p0: IDisplayManagerCallback?) {
+
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt:38]:
```suggestion
```

File: paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  override fun setTemporaryAutoBrightnessAdjustment(p0: Float) {
+    throw UnsupportedOperationException("Minimum IDisplayManager methods are supported.")
+  }
+
+  override fun getMinimumBrightnessCurve(): Curve {
+    throw UnsupportedOperationException("Minimum IDisplayManager methods are supported.")
+  }
+
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziDisplayManager.kt:183]:
```suggestion
```

File: paparazzi/src/test/java/app/cash/paparazzi/PaparazziTest.kt [hunk 6]
```diff
@@ -124,6 +124,24 @@ class PaparazziTest {
...
+        viewTreeObserver.addOnGlobalLayoutListener {
+          onGlobalLayout = true
+        }
+      }
+    }
+
+    paparazzi.snapshot(view)
+
+    assertThat(onGlobalLayout).isTrue
+  }
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/PaparazziTest.kt:143]: nice

**Review by jrodbx — APPROVED** [review]

**Pull request merged**